File: smpl_viewer/projection.py
"""Source-coordinate perspective projection.

Source coords: Y+ = up, -Z = depth (camera at origin looking at -Z).
Formula: u = fx * X / (-Z) + cx, v = fy * (-Y) / (-Z) + cy.
"""
import numpy as np


def project_src(pts, fx, fy, cx, cy):
    """Project source-coordinate points to image pixels.

    Args:
        pts: (N, 3) float array, Y+ up, -Z depth.
        fx, fy, cx, cy: intrinsics in pixels.

    Returns:
        u, v: each (N,) float arrays.

    Raises:
        ValueError: if any point has Z >= 0 (behind/at camera).
    """
    pts = np.asarray(pts, dtype=np.float64)
    X, Y, Z = pts[:, 0], pts[:, 1], pts[:, 2]
    if (Z >= 0).any():
        raise ValueError("points behind camera (Z>=0) cannot be projected in src coords")
    u = fx * X / (-Z) + cx
    v = fy * (-Y) / (-Z) + cy
    return u, v

File: label/src/scene/projection.js
// label/src/scene/projection.js
// Source coords: Y+ up, -Z depth. u = fx*X/(-Z)+cx, v = fy*(-Y)/(-Z)+cy.

export function projectPoint([x, y, z], { fx, fy, cx, cy }) {
  if (z >= 0) throw new Error('point behind camera (Z>=0) cannot be projected');
  return [fx * x / (-z) + cx, fy * (-y) / (-z) + cy];
}

// verts: flat Float32Array [x0,y0,z0, x1,y1,z1, ...]. Returns [x, y, w, h].
export function bboxFromPoints(verts, K) {
  let minU = Infinity;
  let minV = Infinity;
  let maxU = -Infinity;
  let maxV = -Infinity;
  for (let i = 0; i + 2 < verts.length; i += 3) {
    const z = verts[i + 2];
    if (z >= 0) continue;
    const [u, v] = projectPoint([verts[i], verts[i + 1], z], K);
    if (u < minU) minU = u;
    if (v < minV) minV = v;
    if (u > maxU) maxU = u;
    if (v > maxV) maxV = v;
  }
  if (!Number.isFinite(minU)) throw new Error('no projectable points');
  return [minU, minV, maxU - minU, maxV - minV];
}
